### ISSUE TITLE
[Request] Better navigation in a julia buffer.

### ISSUE THREAD

**Issue by 4aHxKzD** [JuliaEditorSupport/julia-emacs#88]
Right now in a `julia-mode` buffer neither `forward-paragraph` nor `forward-sexp` (and their `backward` equivalents) understand julia syntax to move between function definitions and flow-control code blocks. For example, the default `python-mode` offers this.

**Comment by non-Jedi** [comment]
I believe this is a duplicate of #6. Manually defining things like `forward-sex-p-function` seems just as difficult to me as actually building the SMIE parser.

**Comment by tpapp** [comment]
I agree about the difficulty, but I am not closing this as a duplicate since #6 is just one possible solution for this, and until it is implemented it is hard to say whether that is the easiest one.

**Comment by 4aHxKzD** [comment]
I'm curious (sorry if this sounds dumb, but I don't have a CS background), why is it hard to implement a Julia parser in elisp to allow better navigation if the parser in Julia itself is written in a lisp? Couldn't it be adapted?

**Comment by tpapp** [comment]
It could certainly be ported, but note that it is a few thousand lines of code so this is not a trivial effort. Also, to keep it in sync would require constant work.

Instead, I wonder if it would be possible to just maintain a frontend to the FemtoLisp parser that parses and annotates Julia code with metadata ("a function starts here", etc), which would be fast to invoke.

**Comment by 4aHxKzD** [comment]
> It could certainly be ported, but note that it is a few thousand lines of code so this is not a trivial effort. Also, to keep it in sync would require constant work.

So the issue is being a lot of, lets call it, "manual work" instead of intellectual one? I'm asking because imagine that I want to do it myself. I don't know about parsers and stuff like that, but porting code is like translation, so I would not need to know much, right?

**Comment by tpapp** [comment]
It's a different Lisp dialect, and you would have to understand the code I think, so I would imagine that it is a large effort. But who knows, maybe you will find a shortcut.

@JeffBezanson, do you have an opinion on this? Also, do you think it would be possible to use `julia --lisp` to take an input file and parse it into an AST of SEXPs with line number tokens etc, and do it fast? That could solve most of our parsing problems.

**Comment by non-Jedi** [comment]
Note that it won't be strictly mechanical. For one thing, emacs requires you be able to emit the previous token from a position which requires different logic than the femtolisp parser which only parses forward. There's also the mismatch in implementations where the femtolisp parser only has to run through the source and emit an AST which doesn't necessarily retain information about its source code. This is in contrast to something like https://github.com/julia-vscode/CSTParser.jl.

**Comment by tpapp** [comment]
@non-Jedi: I thought @4aHxKzD meant porting [julia-parser.scm](https://github.com/JuliaLang/julia/blob/master/src/julia-parser.scm) *as is* to Emacs Lisp.

**Comment by non-Jedi** [comment]
My point is that doing so is not sufficient to hook into SMIE.

**Comment by non-Jedi** [comment]
As a counterpoint to why porting the parser might be useful, `js2-mode` actually uses an elisp port of the Mozilla Rhino JS parser. Instead of hooking into smie or even the normal `font-lock-defaults`, `js2-mode` parses the file on a timer and manually re-implements syntax highlighting, indentation, and many other rich editing features on top of the parsed CST.

While technically feasible, doing this for julia would be quite a project. `js2-mode` is almost 13,000 lines of code.
